### ISSUE TITLE
Shift + mouse select should work

### ISSUE THREAD

**Issue by daurnimator** [jwilm/alacritty#146]
Carrying over from https://github.com/jwilm/alacritty/issues/6#issuecomment-271057705

To avoid a terminal program from capturing all mouse events, it's common to hold Shift while selecting.

**Comment by AntouanK** [comment]
Installed alacritty on Arch.
Copying with the mouse works normally. 
But inside tmux, I cannot copy with `shift` and the mouse anymore.
Any way to enable that via the config?

**Comment by pmalek** [comment]
I hve the same problem: I use tmux and in other terminals (e.g. Gnome terminal) I used to use Shift + mouse select and then Shift + Ctrl + C to copy but I cannot do with alacritty.

Can this be enabled with config somehow?

**Comment by AntouanK** [comment]
@jwilm Any idea if the copying issue on tmux can/will be solved?

**Comment by jwilm** [comment]
Ohhh I see. So, when shift is held down, Alacritty shouldn't forward events to the pseudoterminal. Seems easy enough to add. The only problem at the moment is glutin doesn't provide modifiers with mouse events at the moment. Otherwise, this should be easy to solve.

**Comment by AntouanK** [comment]
What does that mean though? What is `glutin`?

**Comment by jwilm** [comment]
_glutin_ is the windowing library Alacritty uses. Think GLFW, but in Rust.

**Comment by AntouanK** [comment]
JS/web dev here, no idea about those libraries :zipper_mouth_face: 
But for this issue, does that mean we depend on some change in the `glutin` library that we have no idea if and when it will happen?
I want to use alacritty, but it's getting problematic that I cannot copy stuff. Otherwise it's really good so far.
( great work by the way :+1:  )

**Comment by emersion** [comment]
It seems that this isn't going to happen (https://github.com/tomaka/glutin/pull/87). Maybe alacritty should handle modifiers with another library as tomaka suggested?

**Comment by jwilm** [comment]
@emersion that's a really old issue. I expect once we land modifier forwarding for keyboard events, adding it for mouse events will be a natural extension.

**Comment by nirfse** [comment]
Is there any news on fixing text copying inside "tmux"?

**Comment by ibrokemypie** [comment]
bump

**Comment by algesten** [comment]
@jwilm I don't quite understand why this is an alacritty issue. I was surprised to find any copy-paste functions in alacritty.

For me copy-paste is linked to scrolling since it's quite common to want to copy an area that would require scrolling to reach the boundaries. So if alacritty takes the hard stance on scrolling, delegating it to a multiplexer, why is copy-paste not?

Wouldn't it be cleaner to just drop all this shift-mouse select strangeness?

**Comment by ibrokemypie** [comment]
in various other terminals, holding shift kind of just escapes mouse usage, allowing the terminal emulator to select text rather than the program. shift control c and v already work, but if using tmux the is currently no way to select text to copy out, contrary to other terminals, and also a massive pain

**Comment by algesten** [comment]
@ibrokemypie, yes, I get that, but if you configure your tmux/vi/whatnot to copy to the system buffer, then alacritty doesn't need to do any of this.

**Comment by algesten** [comment]
@ibrokemypie here's my (macOS) [`tmux.conf`](https://gist.github.com/algesten/0ba2d7c19763db6a7396126e8526b65b#file-tmux-conf-L20-L21) with the relevant rows highlighted.

On macOS  the command line tool to put something into the clipboard is `pbcopy`, however due to some annoying appleness this doesn't work straight away, so an extra utility is needed `reattach-to-user-namespace pbcopy`.

For Linux there seems to be a [number of options](https://askubuntu.com/questions/11925/a-command-line-clipboard-copy-and-paste-utility#11927), `xsel`, `xclip`, `clipit`.

I suggest you try out one of them and see if you can get `echo Foo | xclip` from the command line work first. Then slot in whichever worked into your `tmux.conf` in a similar way to mine.

**Comment by algesten** [comment]
Also, if alacritty is using the binding you want to use already, you need to remove that in `alacritty.yml`, since otherwise you don't get the keys sent through to tmux.

**Comment by ibrokemypie** [comment]
is that gist link working, cant seem to access it

**Comment by algesten** [comment]
Should be... https://gist.github.com/algesten/0ba2d7c19763db6a7396126e8526b65b

**Comment by ibrokemypie** [comment]
that looks like it requires switching into different modes to simply copy some text, when it could be as easy as click, drag, ctrl shift c

**Comment by nirfse** [comment]
@algesten 
> if you configure your tmux/vi/whatnot to copy to the system buffer, then alacritty doesn't need to do any of this

What happens, if I have to use remote _tmux_ over _ssh_? Will I still be able to copy/paste text from the remote _tmux_ terminal session to the local system buffer? And if not, what are the most reasonable solutions?

At this point, Alacritty forces me to either use _tmux_ to be able to scroll (ok, I somewhat get author's principle), or not being able to scroll, but to be able to copy. Apart from being annoying, this is the only single reason I still use another terminal on my systems.

**Comment by ibrokemypie** [comment]
@jwilm what are your thoughts? personally I think that its a pretty standard feature and if it isn't coming I will be on the hunt for a terminal emulator again

**Comment by hobarrera** [comment]
[This](https://gitlab.com/hobarrera/dotfiles/blob/master/home/.config/tmux/tmux.conf) works for copying from tmux on linux. Note that no mode switch is requires since mouse will auto-activate it.

I always do shift-drag a lot out a habit though.

**Comment by jwilm** [comment]
@ibrokemypie 
> what are your thoughts? personally I think that its a pretty standard feature

Agree, we should fix it.

**Comment by blark** [comment]
Very glad to hear that this feature is coming. Thanks!

**Comment by sijad** [comment]
this is my tmux config https://gist.github.com/sijad/771426b3995eb05faeb6d78926645e56#file-tmux-conf
working with latest tmux and macOS, just click and drag no need to `ctrl` + `shift` + `c` @ibrokemypie

**Comment by jwilm** [comment]
This is going to be blocked on winit providing modifiers with mouse events.

**Comment by maximbaz** [comment]
Is there a tracking feature request for this to subscribe on winit side? I could find a merged PR that added modifiers to keyboard events (https://github.com/tomaka/winit/pull/112), but nothing for mouse events...

**Comment by maximbaz** [comment]
@nirfse, @hobarrera it is possible to make remote tmux over ssh to copy text into your local clipboard, the trick is to use X11 forwarding and configure tmux to copy with `xclip -selection clipboard` (or use [tmux-plugins/tmux-yank](https://github.com/tmux-plugins/tmux-yank)). See something like [this](https://defuse.ca/blog/clipboard-over-ssh-with-vim.html) for more details.

But the feature is needed, sometimes you just can't control remote server to configure this stuff.

**Comment by hobarrera** [comment]
Yeah, that won't work because xclip isn't generally installed on every other host I ssh into. FWIW, neither is tmux. On the majority of these, I can't even install those things.

**Comment by gilbertw1** [comment]
I've experimented with this and was able to add modifier support to mouse events in winit/glutin. I currently need to test this support on MacOS + Windows and clean up the code a little, but afterwards I'll be submitting a pull request for it.

With that support added to winit, implementing shift-click support is trivial. If you want to test it out, I have a fork branch where shift click works perfectly :)

alacritty w/ shift-click: https://github.com/gilbertw1/alacritty/tree/feature-shift-click
winit mouse modifier support: https://github.com/gilbertw1/winit/commit/9df5042329ead79959c29a89c2aa8d66af1517d2

**Comment by unphased** [comment]
This is the no.1 feature missing that i switch to iTerm for consistently these days. I must use a linux vm or remote linux box sometimes, which means getting the selected buffer from tmux out to system clipboard is impossible.

I am also very glad that it is coming soon.

**Comment by lf94** [comment]
I'm waiting for this too. It has been 1 year. I might go build that other branch linked earlier in this thread.

**Comment by shicks** [comment]
Does anyone know what's blocking this from being merged into the main branch?  This is a pain point I feel daily - I've worked around it clumsily with some tmux and emacs macros, but I'd still be much happier if the terminal could do this natively.

**Comment by AntouanK** [comment]
fwiw, I used @gilbertw1 's fork ( https://github.com/gilbertw1/alacritty/tree/feature-shift-click ) and it works perfectly.
It means I don't get updates for a while, but I wouldn't use alacritty without that feature anyway.

**Comment by gilbertw1** [comment]
@shicks This change depends on functionality in tomaka/winit#328. We won't be able to move forward with the implementation until that PR is merged and modifier support is available with mouse events in winit.

The reason my fork currently works is because it's referencing a fork of glutin/winit that's hosted on github with this functionality included. I'll rebase my branch against master soon for anyone wanting to temporarily use it in the interim.

**Comment by ChinesePeanuts** [comment]
[tomaka/winit#328](https://github.com/tomaka/winit/pull/328) is now merged 🎉 I guess it should be fine now

**Comment by lf94** [comment]
@jwilm now that winit changes are merged, please merge gilbertw1's changes!

**Comment by piahoo** [comment]
thanks @chrisduerr ! now, finally, i can copy and paste in alacritty   :smile:

**Comment by dgmora** [comment]
Sorry for commenting here, I didn't find any other related issue. If I have two splits in tmux, the selection ignores that there are two splits and selects text for both of them. Is this my issue or a general issue? This did not happen to me on tmux + iterm2

**Comment by nirfse** [comment]
@dgmora I can confirm I see the same behavior in my case.

**Comment by pmalek** [comment]
@dgmora @nirfse The same for me. (Linux + alacritty + tmux)
I suggest to open a separate issue for that.

**Comment by svrana** [comment]
Hmm, this fix works for me when mouse mode is off in tmux, but not when it's enabled (i.e., .set -g mouse on). (Linux + alacritty + tmux). Is this expected behavior?

**Comment by dzhang-b** [comment]
I have the problem that when shift+mouse click enabled paste from alacritty.yaml, and I press shift and mouse click to paste in tmux or vim where mouse capture is enabled, the content will be pasted twice. 

Anyone has a solution for this?

**Comment by dzhang-b** [comment]
solution: disable the defualt binding in the config
```
mouse_bindings:
  - { mouse: Middle, mods: Shift, action: PasteSelection }
  - { mouse: Right,  mods: Shift, action: Paste }
  - { mouse: Middle,  action: None }
  - { mouse: Right,  action: None }
```
